### PR TITLE
Disable partial aggregation when size reduction is below threshold

### PR DESCRIPTION
**Pull request by prithvip**
Test plan - 

```
presto:sf1> set session partial_aggregation_strategy='automatic';
SET SESSION
presto:sf1> explain select sum(acctbal) from supplier group by nationkey;
                                                                                        Query Plan                                                                                        
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 - Output[_col0] => [sum:double]                                                                                                                                                          
         _col0 := sum                                                                                                                                                                     
     - RemoteStreamingExchange[GATHER] => [sum:double]                                                                                                                                    
         - Project[projectLocality = LOCAL] => [sum:double]                                                                                                                               
             - Aggregate(FINAL)[nationkey] => [nationkey:bigint, sum:double]                                                                                                              
                     sum := "presto.default.sum"((sum_5))                                                                                                                                 
                 - LocalExchange[HASH][$hashvalue] (nationkey) => [nationkey:bigint, sum_5:double, $hashvalue:bigint]                                                                     
                     - RemoteStreamingExchange[REPARTITION][$hashvalue_6] => [nationkey:bigint, sum_5:double, $hashvalue_6:bigint]                                                        
                         - Project[projectLocality = LOCAL] => [nationkey:bigint, sum_5:double, $hashvalue_7:bigint]                                                                      
                                 $hashvalue_7 := combine_hash(BIGINT 0, COALESCE($operator$hash_code(nationkey), BIGINT 0))                                                               
                             - Aggregate(PARTIAL)[nationkey] => [nationkey:bigint, sum_5:double]                                                                                          
                                     sum_5 := "presto.default.sum"((acctbal))                                                                                                             
                                 - TableScan[TableHandle {connectorId='tpch', connectorHandle='supplier:sf1.0', layout='Optional[supplier:sf1.0]'}] => [nationkey:bigint, acctbal:double] 
                                         Estimates: {rows: 10000 (175.78kB), cpu: 180000.00, memory: 0.00, network: 0.00}                                                                 
                                         nationkey := tpch:nationkey                                                                                                                      
                                         acctbal := tpch:acctbal                                                                                                                          
                                                                                                                                                                                          
(1 row)

Query 20210623_234454_00053_m6nxx, FINISHED, 1 node
http://localhost:8080/ui/query.html?20210623_234454_00053_m6nxx
Splits: 1 total, 1 done (100.00%)
CPU Time: 0.0s total,     0 rows/s,     0B/s, 0% active
Per Node: 0.0 parallelism,     0 rows/s,     0B/s
Parallelism: 0.0
Peak User Memory: 0B
Peak Total Memory: 1.5KB
Peak Task Total Memory: 1.5KB
0:00 [0 rows, 0B] [0 rows/s, 0B/s]

presto:sf1> explain select sum(acctbal) from supplier group by suppkey;
                                                                                                               Query Plan                                                                                                               
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 - Output[_col0] => [sum:double]                                                                                                                                                                                                        
         Estimates: {rows: 10000 (87.89kB), cpu: 1350000.00, memory: 180000.00, network: 360000.00}                                                                                                                                     
         _col0 := sum                                                                                                                                                                                                                   
     - RemoteStreamingExchange[GATHER] => [sum:double]                                                                                                                                                                                  
             Estimates: {rows: 10000 (87.89kB), cpu: 1350000.00, memory: 180000.00, network: 360000.00}                                                                                                                                 
         - Project[projectLocality = LOCAL] => [sum:double]                                                                                                                                                                             
                 Estimates: {rows: 10000 (87.89kB), cpu: 1350000.00, memory: 180000.00, network: 270000.00}                                                                                                                             
             - Aggregate[suppkey] => [suppkey:bigint, sum:double]                                                                                                                                                                       
                     Estimates: {rows: 10000 (175.78kB), cpu: 1260000.00, memory: 180000.00, network: 270000.00}                                                                                                                        
                     sum := "presto.default.sum"((acctbal))                                                                                                                                                                             
                 - LocalExchange[HASH][$hashvalue] (suppkey) => [suppkey:bigint, acctbal:double, $hashvalue:bigint]                                                                                                                     
                         Estimates: {rows: 10000 (263.67kB), cpu: 990000.00, memory: 0.00, network: 270000.00}                                                                                                                          
                     - RemoteStreamingExchange[REPARTITION][$hashvalue_5] => [suppkey:bigint, acctbal:double, $hashvalue_5:bigint]                                                                                                      
                             Estimates: {rows: 10000 (263.67kB), cpu: 720000.00, memory: 0.00, network: 270000.00}                                                                                                                      
                         - ScanProject[table = TableHandle {connectorId='tpch', connectorHandle='supplier:sf1.0', layout='Optional[supplier:sf1.0]'}, projectLocality = LOCAL] => [suppkey:bigint, acctbal:double, $hashvalue_6:bigint] 
                                 Estimates: {rows: 10000 (263.67kB), cpu: 180000.00, memory: 0.00, network: 0.00}/{rows: 10000 (263.67kB), cpu: 450000.00, memory: 0.00, network: 0.00}                                                 
                                 $hashvalue_6 := combine_hash(BIGINT 0, COALESCE($operator$hash_code(suppkey), BIGINT 0))                                                                                                               
                                 suppkey := tpch:suppkey                                                                                                                                                                                
                                 acctbal := tpch:acctbal                                                                                                                                                                                
                                                                                                                                                                                                                                        
(1 row)

Query 20210623_234504_00054_m6nxx, FINISHED, 1 node
http://localhost:8080/ui/query.html?20210623_234504_00054_m6nxx
Splits: 1 total, 1 done (100.00%)
CPU Time: 0.0s total,     0 rows/s,     0B/s, 0% active
Per Node: 0.0 parallelism,     0 rows/s,     0B/s
Parallelism: 0.0
Peak User Memory: 0B
Peak Total Memory: 1.92KB
Peak Task Total Memory: 1.92KB
0:00 [0 rows, 0B] [0 rows/s, 0B/s]

presto:sf1> explain select sum(acctbal) from supplier group by suppkey, nationkey;
                                                                                                                          Query Plan                                                                                                                          
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 - Output[_col0] => [sum:double]                                                                                                                                                                                                                              
         _col0 := sum                                                                                                                                                                                                                                         
     - RemoteStreamingExchange[GATHER] => [sum:double]                                                                                                                                                                                                        
         - Project[projectLocality = LOCAL] => [sum:double]                                                                                                                                                                                                   
             - Aggregate(FINAL)[suppkey, nationkey][$hashvalue] => [suppkey:bigint, nationkey:bigint, $hashvalue:bigint, sum:double]                                                                                                                          
                     sum := "presto.default.sum"((sum_6))                                                                                                                                                                                                     
                 - LocalExchange[HASH][$hashvalue] (suppkey, nationkey) => [suppkey:bigint, nationkey:bigint, sum_6:double, $hashvalue:bigint]                                                                                                                
                     - RemoteStreamingExchange[REPARTITION][$hashvalue_7] => [suppkey:bigint, nationkey:bigint, sum_6:double, $hashvalue_7:bigint]                                                                                                            
                         - Aggregate(PARTIAL)[suppkey, nationkey][$hashvalue_8] => [suppkey:bigint, nationkey:bigint, $hashvalue_8:bigint, sum_6:double]                                                                                                      
                                 sum_6 := "presto.default.sum"((acctbal))                                                                                                                                                                                     
                             - ScanProject[table = TableHandle {connectorId='tpch', connectorHandle='supplier:sf1.0', layout='Optional[supplier:sf1.0]'}, projectLocality = LOCAL] => [suppkey:bigint, nationkey:bigint, acctbal:double, $hashvalue_8:bigint] 
                                     Estimates: {rows: 10000 (351.56kB), cpu: 270000.00, memory: 0.00, network: 0.00}/{rows: 10000 (351.56kB), cpu: 630000.00, memory: 0.00, network: 0.00}                                                                   
                                     $hashvalue_8 := combine_hash(combine_hash(BIGINT 0, COALESCE($operator$hash_code(suppkey), BIGINT 0)), COALESCE($operator$hash_code(nationkey), BIGINT 0))                                                               
                                     suppkey := tpch:suppkey                                                                                                                                                                                                  
                                     acctbal := tpch:acctbal                                                                                                                                                                                                  
                                     nationkey := tpch:nationkey                                                                                                                                                                                              
                                                                                                                                                                                                                                                              
(1 row)

Query 20210623_234512_00055_m6nxx, FINISHED, 1 node
http://localhost:8080/ui/query.html?20210623_234512_00055_m6nxx
Splits: 1 total, 1 done (100.00%)
CPU Time: 0.0s total,     0 rows/s,     0B/s, 0% active
Per Node: 0.0 parallelism,     0 rows/s,     0B/s
Parallelism: 0.0
Peak User Memory: 0B
Peak Total Memory: 1.78KB
Peak Task Total Memory: 1.78KB
0:00 [0 rows, 0B] [0 rows/s, 0B/s]

presto:sf1> explain select sum(acctbal) from supplier group by suppkey * 2;
                                                                                                       Query Plan                                                                                                       
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 - Output[_col0] => [sum:double]                                                                                                                                                                                        
         _col0 := sum                                                                                                                                                                                                   
     - RemoteStreamingExchange[GATHER] => [sum:double]                                                                                                                                                                  
         - Project[projectLocality = LOCAL] => [sum:double]                                                                                                                                                             
             - Aggregate(FINAL)[expr] => [expr:bigint, sum:double]                                                                                                                                                      
                     sum := "presto.default.sum"((sum_5))                                                                                                                                                               
                 - LocalExchange[HASH][$hashvalue] (expr) => [expr:bigint, sum_5:double, $hashvalue:bigint]                                                                                                             
                     - RemoteStreamingExchange[REPARTITION][$hashvalue_6] => [expr:bigint, sum_5:double, $hashvalue_6:bigint]                                                                                           
                         - Project[projectLocality = LOCAL] => [expr:bigint, sum_5:double, $hashvalue_7:bigint]                                                                                                         
                                 $hashvalue_7 := combine_hash(BIGINT 0, COALESCE($operator$hash_code(expr), BIGINT 0))                                                                                                  
                             - Aggregate(PARTIAL)[expr] => [expr:bigint, sum_5:double]                                                                                                                                  
                                     sum_5 := "presto.default.sum"((acctbal))                                                                                                                                           
                                 - ScanProject[table = TableHandle {connectorId='tpch', connectorHandle='supplier:sf1.0', layout='Optional[supplier:sf1.0]'}, projectLocality = LOCAL] => [expr:bigint, acctbal:double] 
                                         Estimates: {rows: 10000 (175.78kB), cpu: 180000.00, memory: 0.00, network: 0.00}/{rows: 10000 (175.78kB), cpu: 360000.00, memory: 0.00, network: 0.00}                         
                                         expr := (suppkey) * (BIGINT 2)                                                                                                                                                 
                                         suppkey := tpch:suppkey                                                                                                                                                        
                                         acctbal := tpch:acctbal                                                                                                                                                        
                                                                                                                                                                                                                        
(1 row)

Query 20210623_234528_00056_m6nxx, FINISHED, 1 node
http://localhost:8080/ui/query.html?20210623_234528_00056_m6nxx
Splits: 1 total, 1 done (100.00%)
CPU Time: 0.0s total,     0 rows/s,     0B/s, 0% active
Per Node: 0.0 parallelism,     0 rows/s,     0B/s
Parallelism: 0.0
Peak User Memory: 0B
Peak Total Memory: 1.62KB
Peak Task Total Memory: 1.62KB
0:00 [0 rows, 0B] [0 rows/s, 0B/s]
```

```
== NO RELEASE NOTE ==
```

### CODE REVIEW

**Review by ahmadghazal64 — CHANGES_REQUESTED** [review]

File: presto-main/src/main/java/com/facebook/presto/cost/ProjectStatsRule.java [hunk 1]
```diff
@@ -53,7 +53,8 @@ public ProjectStatsRule(ScalarStatsCalculator scalarStatsCalculator, StatsNormal
     {
         PlanNodeStatsEstimate sourceStats = statsProvider.getStats(node.getSource());
         PlanNodeStatsEstimate.Builder calculatedStats = PlanNodeStatsEstimate.builder()
-                .setOutputRowCount(sourceStats.getOutputRowCount());
+                .setOutputRowCount(sourceStats.getOutputRowCount())
+                .setConfident(sourceStats.isConfident());
```
Comment (arhimondr) [inline presto-main/src/main/java/com/facebook/presto/cost/ProjectStatsRule.java:57]: Based on the offline discussion: Let's set it to confident only when we are confident no columns have changed the cardinallity

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushPartialAggregationThroughExchange.java [hunk 2]
```diff
@@ -265,6 +275,17 @@ private PlanNode split(AggregationNode node, Context context)
                 node.getGroupIdVariable());
     }
 
+    private boolean byteReductionLessThanThreshold(AggregationNode aggregationNode, ExchangeNode exchangeNode, Context context)
+    {
+        StatsProvider stats = context.getStatsProvider();
+        PlanNodeStatsEstimate exchangeStats = stats.getStats(exchangeNode);
+        PlanNodeStatsEstimate aggregationStats = stats.getStats(aggregationNode);
```
Comment (arhimondr) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushPartialAggregationThroughExchange.java:282]: This estimate might not be reliable, especially when aggregating on more than one column. What do you think about looking at the `exchangeStats` only? Can we extract the cardinallity for the aggregation column (if there's only a single column) and check if the cardinality above a threshold?

File: presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestPushPartialAggregationThroughExchange.java [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+                                            .singleDistributionPartitioningScheme(a)))
+                            .addAggregation(p.variable("SUM", DOUBLE), expression("SUM(a)"), ImmutableList.of(DOUBLE))
+                            .globalGrouping()
+                            .step(PARTIAL));
+                })
+                .doesNotFire();
+    }
+
+    @Test
+    public void testNoPartialAggregationWhenReductionBelowThreshold()
```
[the remark this answers is not in the document]
Comment (prithvip) [reply presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestPushPartialAggregationThroughExchange.java:86]: Added a test to make sure we don't disable the partial agg when we are not confident

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushPartialAggregationThroughExchange.java [hunk 4]
```diff
@@ -265,6 +276,17 @@ private PlanNode split(AggregationNode node, Context context)
                 node.getGroupIdVariable());
     }
 
+    private boolean byteReductionLessThanThreshold(AggregationNode aggregationNode, ExchangeNode exchangeNode, Context context)
```
[the remark this answers is not in the document]
Comment (prithvip) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushPartialAggregationThroughExchange.java]: Changed to partialAggregationNotUseful

File: presto-main/src/main/java/com/facebook/presto/cost/ProjectStatsRule.java (hunk 1 above)
Comment (prithvip) [reply presto-main/src/main/java/com/facebook/presto/cost/ProjectStatsRule.java]: Done

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushPartialAggregationThroughExchange.java (hunk 2 above)
Comment (prithvip) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushPartialAggregationThroughExchange.java:284]: Limited the scope of disabling partial aggregation to group bys on a single column

**Review by arhimondr — APPROVED** [review]

**Pull request merged**